# Consigna del desafío 1

**1**. Vectorizar documentos. Tomar 5 documentos al azar y medir similaridad con el resto de los documentos.
Estudiar los 5 documentos más similares de cada uno analizar si tiene sentido
la similaridad según el contenido del texto y la etiqueta de clasificación.

**2**. Entrenar modelos de clasificación Naïve Bayes para maximizar el desempeño de clasificación
(f1-score macro) en el conjunto de datos de test. Considerar cambiar parámteros
de instanciación del vectorizador y los modelos y probar modelos de Naïve Bayes Multinomial
y ComplementNB.

**3**. Transponer la matriz documento-término. De esa manera se obtiene una matriz
término-documento que puede ser interpretada como una colección de vectorización de palabras.
Estudiar ahora similaridad entre palabras tomando 5 palabras y estudiando sus 5 más similares. **La elección de palabras no debe ser al azar para evitar la aparición de términos poco interpretables, elegirlas "manualmente"**.

## Imports y preparación dataset

In [136]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.naive_bayes import MultinomialNB, ComplementNB
from sklearn.metrics import f1_score
from sklearn.datasets import fetch_20newsgroups
import numpy as np
import random

In [137]:
newsgroups_train = fetch_20newsgroups(subset='train', remove=('headers', 'footers', 'quotes'))
newsgroups_test = fetch_20newsgroups(subset='test', remove=('headers', 'footers', 'quotes'))
display(newsgroups_test.target_names)

tfidfvect = TfidfVectorizer()
X_train = tfidfvect.fit_transform(newsgroups_train.data)
idx2word = {v: k for k,v in tfidfvect.vocabulary_.items()}
y_train = newsgroups_train.target

['alt.atheism',
 'comp.graphics',
 'comp.os.ms-windows.misc',
 'comp.sys.ibm.pc.hardware',
 'comp.sys.mac.hardware',
 'comp.windows.x',
 'misc.forsale',
 'rec.autos',
 'rec.motorcycles',
 'rec.sport.baseball',
 'rec.sport.hockey',
 'sci.crypt',
 'sci.electronics',
 'sci.med',
 'sci.space',
 'soc.religion.christian',
 'talk.politics.guns',
 'talk.politics.mideast',
 'talk.politics.misc',
 'talk.religion.misc']

### Ejemplo similitud de documentos 

In [ ]:
idx = 4811
display(newsgroups_train.data[idx][:200])
cossim = cosine_similarity(X_train[idx], X_train)[0]
np.sort(cossim)[::-1]       #Valores de similaridad ordenados de mayor a menor
np.argsort(cossim)[::-1]    #Indices de documentos similares de mayor a menor
mostsim = np.argsort(cossim)[::-1][1:6] #Top 5
for i in mostsim:
  display(newsgroups_train.target_names[y_train[i]])

"THE WHITE HOUSE\n\n                  Office of the Press Secretary\n                   (Pittsburgh, Pennslyvania)\n______________________________________________________________\nFor Immediate Release                         April 17, 1993     \n\n             \n                  RADIO ADDRESS TO THE NATION \n                        BY THE PRESIDENT\n             \n                Pittsburgh International Airport\n                    Pittsburgh, Pennsylvania\n             \n             \n10:06 A.M. EDT\n             \n             \n             THE PRESIDENT:  Good morning.  My voice is coming to\nyou this morning through the facilities of the oldest radio\nstation in America, KDKA in Pittsburgh.  I'm visiting the city to\nmeet personally with citizens here to discuss my plans for jobs,\nhealth care and the economy.  But I wanted first to do my weekly\nbroadcast with the American people. \n             \n             I'm told this station first broadcast in 1920 when\nit reported th

'talk.politics.misc'

'talk.politics.misc'

'talk.politics.misc'

'talk.politics.misc'

'talk.politics.misc'

### Ejemplo clasificación

In [139]:
clf = MultinomialNB()
clf.fit(X_train, y_train)
X_test = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target
y_pred =  clf.predict(X_test)
display(f"F1 score {f1_score(y_test, y_pred, average='macro')}")

'F1 score 0.5854345727938506'

## Desarrollo 

### Punto 1

In [140]:
def get_docs_cosine_similarity(X_train, idx, n):
    topn_dict = {"doc_idx": idx}
    cossim = cosine_similarity(X_train[idx], X_train)[0]    
    topn_dict.update({"docs_sim_idx" : np.argsort(cossim)[::-1][1:n+1], "docs_sim_cosin" : np.sort(cossim)[::-1][1:n+1]})
    return topn_dict

def print_doc_info(dataset, topn_docs):
    for docs in topn_docs:
        text = dataset.data[docs['doc_idx']]
        type = dataset.target_names[int(dataset.target[docs['doc_idx']])]
        display(f"Doc referencia ID:{docs['doc_idx']}, clase: {type}")
        display(f"CONTENIDO: {text[:200]}")
        for i in range(len(docs['docs_sim_idx'])):
            text = dataset.data[docs['docs_sim_idx'][i]]
            type = dataset.target_names[int(dataset.target[docs['docs_sim_idx'][i]])]
            display(f"{i+1}) Doc similar ID: {docs['docs_sim_idx'][i]}, similitud coseno: {docs['docs_sim_cosin'][i]}, clase: {type}")
            display(f"CONTENIDO: {text[:200]}")
        display('-'*100)

In [141]:
seed = 42
random.seed(seed)
n_docs = 5
sample_docs_idx = random.sample(range(X_train.shape[0]), n_docs)

topn_docs = []
for i in sample_docs_idx:
    topn_docs.append(get_docs_cosine_similarity(X_train, i, 5))
print_doc_info(newsgroups_train, topn_docs)


'Doc referencia ID:10476, clase: rec.sport.hockey'

'CONTENIDO: This is a general question for US readers:\n\nHow extensive is the playoff coverage down there?  In Canada, it is almost\nimpossible not to watch a series on TV (ie the only two series I have not had\nan '

'1) Doc similar ID: 5064, similitud coseno: 0.22503645427838304, clase: rec.sport.hockey'

"CONTENIDO: \nI only have one comment on this:  You call this a *classic* playoff year\nand yet you don't include a Chicago-Detroit series.  C'mon, I'm a Boston\nfan and I even realize that Chicago-Detroit games are"

'2) Doc similar ID: 9623, similitud coseno: 0.21743229780919432, clase: talk.politics.mideast'

'CONTENIDO: Accounts of Anti-Armenian Human Right Violations in Azerbaijan #012\n                 Prelude to Current Events in Nagorno-Karabakh\n\n        +---------------------------------------------------------+\n'

'3) Doc similar ID: 10575, similitud coseno: 0.21644447761221192, clase: sci.crypt'

'CONTENIDO: \nI am not an expert in the cryptography science, but some basic things\nseem evident to me, things which this Clinton Clipper do not address.\nThe all pertain to opportunites for abuse, and conclusions '

'4) Doc similar ID: 10836, similitud coseno: 0.21260319521108376, clase: alt.atheism'

'CONTENIDO: Archive-name: atheism/faq\nAlt-atheism-archive-name: faq\nLast-modified: 5 April 1993\nVersion: 1.1\n\n                    Alt.Atheism Frequently-Asked Questions\n\nThis file contains responses to articles w'

'5) Doc similar ID: 2350, similitud coseno: 0.21107117783786433, clase: sci.crypt'

'CONTENIDO: Archive-name: net-privacy/part1\nLast-modified: 1993/3/3\nVersion: 2.1\n\n\nIDENTITY, PRIVACY, and ANONYMITY on the INTERNET\n================================================\n\n(c) 1993 L. Detweiler.  Not fo'

'----------------------------------------------------------------------------------------------------'

'Doc referencia ID:1824, clase: comp.sys.mac.hardware'

'CONTENIDO: \n\n\tI think this kind of comparison is pretty useless in general.  The\nprocessor is only good when a good computer is designed around it adn the\ncomputer is used in its designed purpose.  Comparing pro'

'1) Doc similar ID: 9921, similitud coseno: 0.35419792160817193, clase: comp.sys.mac.hardware'

'CONTENIDO: dhk@ubbpc.uucp (Dave Kitabjian) writes ...\n\n040 486 030 386 020 286\n\n\n060 fastest, then Pentium, with the first versions of the PowerPC\nsomewhere in the vicinity.\n\n\nNo.  Computer speed is only partly '

'2) Doc similar ID: 6364, similitud coseno: 0.31319104088093985, clase: comp.sys.mac.hardware'

'CONTENIDO: Well folks, after some thought the answer struck me flat in the face:\n\n"Why would Apple release a Duo Dock with a processor of its own?"\n\nHere\'s why- People have hounded Apple for a notebook with a 68'

'3) Doc similar ID: 5509, similitud coseno: 0.30413670525282677, clase: comp.sys.mac.hardware'

'CONTENIDO: rvenkate@ux4.cso.uiuc.edu (Ravikuma Venkateswar) writes ...\n\nBenchmarks are for marketing dweebs and CPU envy.  OK, if it will make\nyou happy, the 486 is faster than the 040.  BFD.  Both architectures'

'4) Doc similar ID: 2641, similitud coseno: 0.2503826782386589, clase: comp.sys.mac.hardware'

'CONTENIDO: \n    I think this is mostly the fault of the people who write up the\nliterature and price lists being confused themselves. Since there are\ntwo possible processor configurations and one of the them doe'

'5) Doc similar ID: 4359, similitud coseno: 0.24167619736986773, clase: comp.sys.mac.hardware'

'CONTENIDO: If you get the Centris 650 with CD configuration, you are getting a Mac with\na 68RC040 processor that has built-in math coprocessor support.  My \nunderstanding is that the "optional fpu" refers to you'

'----------------------------------------------------------------------------------------------------'

'Doc referencia ID:409, clase: comp.graphics'

"CONTENIDO: I can't fiqure this out.  I have properly compiled pov on a unix machine\nrunning SunOS 4.1.3  The problem is that when I run the sample .pov files and\nuse the EXACT same parameters when compiling diff"

'1) Doc similar ID: 3444, similitud coseno: 0.23053284689766265, clase: comp.graphics'

"CONTENIDO: Hi, I'm just getting into PoVRay and I was wondering if there is a graphic\npackage that outputs .POV files.  Any help would be appreciated.\nThanks.\n\nLater'ish\nCraig\n"

'2) Doc similar ID: 5799, similitud coseno: 0.20907273908737237, clase: comp.graphics'

'CONTENIDO: I finally got a 24 bit viewer for my POVRAY generated .TGA files.\n\nIt was written in C by Sean Malloy and he kindly sent me a copy.  He\nwrote it for the same purpose, to view .TGA files using his Spee'

'3) Doc similar ID: 5905, similitud coseno: 0.1981754261808808, clase: comp.graphics'

"CONTENIDO: \nHallo POV-Renderers !\nI've got a BocaX3 Card. Now I try to get POV displaying True Colors\nwhile rendering. I've tried most of the options and UNIVESA-Driver\nbut what happens isn't correct.\nCan anybod"

'4) Doc similar ID: 1764, similitud coseno: 0.18384984554054173, clase: comp.graphics'

"CONTENIDO: hi guys\n like all people in this group i'm a fans of fractal and render sw\n my favourite are fractint pov & 3dstudio 2.0 \n now listen my ideas\n i'have just starting now to be able to use 3dstudio quit"

'5) Doc similar ID: 3364, similitud coseno: 0.16585392152558243, clase: comp.graphics'

'CONTENIDO: Does anyone know of a good way (standard PC application/PD utility) to\nconvert tif/img/tga files into LaserJet III format.  We would also like to\ndo the same, converting to HPGL (HP plotter) files.\n\nP'

'----------------------------------------------------------------------------------------------------'

'Doc referencia ID:4506, clase: rec.autos'

"CONTENIDO: \nThis does sound good, but I heard it tends to leave more grit, etc in the \noil pan.  Also, I've been told to change the old when it's hot before the\ngrit has much time to settle.\n\nAny opinions?\n"

'1) Doc similar ID: 4211, similitud coseno: 0.18936063539057416, clase: rec.motorcycles'

"CONTENIDO: \n\nIt's normal for the BMW K bikes to use a little oil in the first few thousand \nmiles.  I don't know why.  I've had three new K bikes, and all three used a\nbit of oil when new - max maybe .4 quart in"

'2) Doc similar ID: 5928, similitud coseno: 0.16823674420892193, clase: comp.sys.mac.hardware'

'CONTENIDO: or\nthere\n\n\nOkay, I guess its time for a quick explanation of Mac sound.\n\nThe original documentation for the sound hardware (IM-3) documents how to\nmake sound by directly accessing hardware.  Basically'

'3) Doc similar ID: 6224, similitud coseno: 0.1583340724404407, clase: rec.autos'

'CONTENIDO: Archive-name: rec-autos/part5\n\n[this article is one of a pair of articles containing commonly\nasked automotive questions; the other article contains questions\nmore geared to the automotive enthusiast,'

'4) Doc similar ID: 5171, similitud coseno: 0.157693767056016, clase: rec.autos'

"CONTENIDO: \nIf the tire has a leak you should fix it. \n\n\nDoesn't work too well if the engine is hot, its more accurate to check the\noil when the engine is cool, i.e. not when you are at a gas station."

'5) Doc similar ID: 9491, similitud coseno: 0.1521899666815704, clase: rec.autos'

'CONTENIDO: My friend brought a subaru SVX recently.  I had drove it for couples times and I\nthink its a great car, esp on snow.  However when she took it to a local Subaru\ndealer for a oil change, the bill came '

'----------------------------------------------------------------------------------------------------'

'Doc referencia ID:4012, clase: rec.sport.hockey'

"CONTENIDO: For those Leaf fans who are concerned, the following players are slated for\nreturn on Thursday's Winnipeg-Toronto game :\n    Peter Zezel, John Cullen\n\n  Mark Osborne and Dave Ellett are questionable t"

'1) Doc similar ID: 6599, similitud coseno: 0.16001317417336078, clase: soc.religion.christian'

'CONTENIDO: True.\n\nAlso read 2 Peter 3:16\n\nPeter warns that the scriptures are often hard to understand by those who\nare not learned on the subject.'

'2) Doc similar ID: 10644, similitud coseno: 0.14276898565096252, clase: rec.sport.hockey'

"CONTENIDO: In  <1qvos8$r78@cl.msu.>, vergolin@euler.lbs.msu.edu (David Vergolini) writes...\n\nThere's quite a few Wings fans lurking about here, they just tend\nto be low key and thoughtful rather than woofers.  I"

'3) Doc similar ID: 7478, similitud coseno: 0.13582073511489307, clase: rec.sport.hockey'

'CONTENIDO: Toronto                          1 1 1--3\nDetroit                          1 4 1--6\nFirst period\n     1, Detroit, Yzerman 1 (Gallant, Ciccarelli) 4:48.\n     2, Toronto, Cullen 1 (Clark, Gill) 10:44.\nS'

'4) Doc similar ID: 7308, similitud coseno: 0.13177395037929182, clase: rec.sport.hockey'

"CONTENIDO: Detroit is a very disciplined team.  There's a lot of Europeans\nin Detroit which would make the game fast, so Toronto would have\nto slow the game down, which means drawing penalties, as a last\nresort "

'5) Doc similar ID: 10792, similitud coseno: 0.13073210977865626, clase: rec.sport.baseball'

'CONTENIDO: \n\nThe tribe will be in town from April 16 to the 19th.\nThere are ALWAYS tickets available! (Though they are playing Toronto,\nand many Toronto fans make the trip to Cleveland as it is easier to\nget tic'

'----------------------------------------------------------------------------------------------------'

Los resultados obtenidos indican que la similitud coseno aplicada a vectores TF-IDF es una herramienta útil para identificar documentos con contenido relacionado al texto de referencia, tanto desde el punto de vista semántico como categórico siempre y cuando exista una alta correlación. 

Cuando los valores de similitud son elevados, es común que los documentos recuperados pertenezcan a la misma clase y compartan términos específicos del tema. No obstante, en situaciones donde la similitud medida es baja, los documentos obtenidos tienden a no guardar una conexión semántica clara ni coincidir en categoría, lo que sugiere que la vectorización no logra reflejar adecuadamente el contenido textual.

### Punto 2

In [142]:
def cls_test (X_train_tfidf, X_test_tfidf):
    clf = MultinomialNB()
    clf.fit(X_train_tfidf, y_train)
    y_pred =  clf.predict(X_test_tfidf)
    display(f"F1 score para MultinomialNB {f1_score(y_test, y_pred, average='macro')}")

    clf = ComplementNB()
    clf.fit(X_train_tfidf, y_train)
    y_pred =  clf.predict(X_test_tfidf)
    display(f"F1 score para ComplementNB {f1_score(y_test, y_pred, average='macro')}")

In [143]:
newsgroups_train = fetch_20newsgroups(subset='train', remove=('headers', 'footers', 'quotes'))
newsgroups_test = fetch_20newsgroups(subset='test', remove=('headers', 'footers', 'quotes'))

tfidfvect = TfidfVectorizer()
X_train_tfidf = tfidfvect.fit_transform(newsgroups_train.data)
idx2word = {v: k for k,v in tfidfvect.vocabulary_.items()}
y_train = newsgroups_train.target
X_test_tfidf = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target

display("TF-IDF default")
cls_test (X_train_tfidf, X_test_tfidf)

tfidfvect = TfidfVectorizer(stop_words='english', ngram_range=(1,2), max_df=0.9, min_df=2)
X_train_tfidf = tfidfvect.fit_transform(newsgroups_train.data)
idx2word = {v: k for k,v in tfidfvect.vocabulary_.items()}
y_train = newsgroups_train.target
X_test_tfidf = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target

display("TF-IDF con parámetros")
cls_test (X_train_tfidf, X_test_tfidf)

'TF-IDF default'

'F1 score para MultinomialNB 0.5854345727938506'

'F1 score para ComplementNB 0.692953349950875'

'TF-IDF con parámetros'

'F1 score para MultinomialNB 0.6503941723196313'

'F1 score para ComplementNB 0.7037173109758078'

Se compararon los modelos MultinomialNB y ComplementNB utilizando representaciones TF-IDF, tanto con la configuración por defecto como con parámetros ajustados (eliminación de stop words en inglés, uso de n-gramas y filtros de frecuencia). 

Esta diferencia se debe a que ComplementNB fue diseñado como una mejora del MultinomialNB para manejar mejor clases desbalanceadas o términos poco representativos, lo que lo vuelve más robusto en configuraciones básicas. Sin embargo, al aplicar una vectorización más cuidada, MultinomialNB mejora notablemente (de 0.585 a 0.650), reduciendo la brecha de desempeño con ComplementNB, que también mejora ligeramente a 0.704.

Estos resultados muestran que ComplementNB es más tolerante a configuraciones simples, mientras que MultinomialNB se beneficia significativamente de una selección adecuada de características.

### Punto 3

In [144]:
def get_words_cosine_similarity(X_train_T, tfidfvect, word, n):
    topn_dict = {"word": word}
    try:
        word_idx = tfidfvect.vocabulary_[word]
    except:
        print(f"{word} no in vocabulary")
    topn_dict.update({"word_idx": word_idx})
    cossim = cosine_similarity(X_train_T[word_idx], X_train_T)[0]
    topn_dict.update({"word_sim_idx" : np.argsort(cossim)[::-1][1:n+1], "word_sim_cosin" : np.sort(cossim)[::-1][1:n+1]})
    return topn_dict

def print_word_info(idx2word, topn_words):
    for words in topn_words:
        display(f"Palabra referencia {words['word']}, ID:{words['word_idx']}")
        for i in range(len(words['word_sim_idx'])):
            display(f"{i+1}) Palabra similar {idx2word[words['word_sim_idx'][i]]}, ID: {words['word_sim_idx'][i]}, similitud coseno: {words['word_sim_cosin'][i]}")
        display('-'*100)

In [145]:
tfidfvect = TfidfVectorizer(stop_words='english', max_df=0.9, min_df=2) #No voy a utilizar ngrams para que no me busque por pares de palabras
X_train_tfidf = tfidfvect.fit_transform(newsgroups_train.data)
idx2word = {v: k for k,v in tfidfvect.vocabulary_.items()}
y_train = newsgroups_train.target

X_train_T = X_train_tfidf.transpose()
words2test = ['motorcycle', 'cpu', 'football', 'atheist', 'toronto']
topn_words = []
for word in words2test:
    topn_words.append(get_words_cosine_similarity(X_train_T, tfidfvect, word, 5))
print_word_info(idx2word, topn_words)

'Palabra referencia motorcycle, ID:23793'

'1) Palabra similar wheelie, ID: 37637, similitud coseno: 0.404315501394809'

'2) Palabra similar shaft, ID: 31561, similitud coseno: 0.2692545582932047'

'3) Palabra similar learner, ID: 21091, similitud coseno: 0.23783674199976998'

'4) Palabra similar bluntly, ID: 7229, similitud coseno: 0.2302796637099688'

'5) Palabra similar enthusiast, ID: 13800, similitud coseno: 0.2256922997195348'

'----------------------------------------------------------------------------------------------------'

'Palabra referencia cpu, ID:10550'

'1) Palabra similar dislodge, ID: 12367, similitud coseno: 0.356190268602508'

'2) Palabra similar heatsink, ID: 17394, similitud coseno: 0.3086652738160754'

'3) Palabra similar cooler, ID: 10292, similitud coseno: 0.29130050924035983'

'4) Palabra similar bending, ID: 6810, similitud coseno: 0.25026188658596993'

'5) Palabra similar tasking, ID: 34291, similitud coseno: 0.2498389957627389'

'----------------------------------------------------------------------------------------------------'

'Palabra referencia football, ID:15412'

'1) Palabra similar madden, ID: 22094, similitud coseno: 0.34695776542107865'

'2) Palabra similar mich, ID: 23097, similitud coseno: 0.304655737958004'

'3) Palabra similar zabriskie, ID: 38913, similitud coseno: 0.29979462178249405'

'4) Palabra similar basketball, ID: 6554, similitud coseno: 0.2933647593152447'

'5) Palabra similar sportstalk, ID: 32810, similitud coseno: 0.28860827446505277'

'----------------------------------------------------------------------------------------------------'

'Palabra referencia atheist, ID:5889'

'1) Palabra similar atheism, ID: 5888, similitud coseno: 0.33245031466909036'

'2) Palabra similar anarchal, ID: 5086, similitud coseno: 0.31034751316326226'

'3) Palabra similar atheists, ID: 5891, similitud coseno: 0.2947785138421634'

'4) Palabra similar doubtless, ID: 12747, similitud coseno: 0.29430041416345676'

'5) Palabra similar techno, ID: 34379, similitud coseno: 0.2861931096814526'

'----------------------------------------------------------------------------------------------------'

'Palabra referencia toronto, ID:35041'

'1) Palabra similar detroit, ID: 11930, similitud coseno: 0.3812332304163351'

'2) Palabra similar gilmour, ID: 16314, similitud coseno: 0.2479111071502925'

'3) Palabra similar chicago, ID: 8879, similitud coseno: 0.24710485460503068'

'4) Palabra similar eecg, ID: 13301, similitud coseno: 0.19442096527109395'

'5) Palabra similar disciplined, ID: 12269, similitud coseno: 0.1834678555585414'

'----------------------------------------------------------------------------------------------------'

Transponer la matriz TF-IDF de documentos genera una matriz que representa términos en función de los documentos en los que aparecen. El significado de cada término queda definido por su distribución a lo largo del corpus, lo cual está limitado por la cantidad y la categoría de los documentos disponibles.

En los ejemplos seleccionados manualmente se observa que términos como 'motorcycle' están fuertemente asociados a 'wheelie' y 'shaft', o que 'cpu' se relaciona con 'heatsink', 'cooler' o 'tasking'. De forma similar, Toronto muestra correlación con otras ciudades como Detroit o Chicago. Sin embargo, el método no es completamente preciso, ya que algunas asociaciones carecen de sentido aparente, posiblemente porque las relaciones entre palabras se manifiestan a nivel de documento y no de coocurrencia directa.

TF-IDF este método no considera contexto, orden de palabras ni sintaxis. Para mejorar la representación semántica de las palabras, se recurre a técnicas más avanzadas como los word embeddings, que capturan relaciones contextuales de forma más robusta.